src/irmin-pack/pack_store.ml

```diff
@@ -166,39 +178,49 @@ module Maker
       let buf = Bytes.create len in
       let n = IO.read t.pack.block ~off buf in
       if n <> len then raise Invalid_read;
-      let hash off = io_read_and_decode_hash ~off t in
+      let hash off =
+        match Offsets.find t.offsets off with
+        | v -> Val.hash v
+        | exception Not_found -> io_read_and_decode_hash ~off t
+      in
```

Comment:
That `hash` function is called from `decode_bin`, on the children of the value that is being decoded.

It is extremely unlikely that `Offsets.find t.offsets off` results in a cache hit because these children are being seen for the first time